[CASE missing_settings_ignored_flag]
[env MYPY_DJANGO_CONFIG=${MYPY_CWD}/mypy_django.ini]
[disable_cache]
from django.conf import settings
reveal_type(settings.NO_SUCH_SETTING)  # N: Revealed type is 'Any'

[file mypy_django.ini]
[[mypy_django_plugin]
ignore_missing_settings = True
[/CASE]

[CASE django_settings_via_config_file]
[env MYPY_DJANGO_CONFIG=${MYPY_CWD}/mypy_django.ini]
[disable_cache]
from django.conf import settings
reveal_type(settings.MY_SETTING)  # N: Revealed type is 'builtins.int'

[file mypy_django.ini]
[[mypy_django_plugin]
django_settings = mysettings

[file mysettings.py]
MY_SETTING: int = 1
[/CASE]

[CASE mypy_django_ini_in_current_directory_is_a_default]
[disable_cache]
from django.conf import settings
reveal_type(settings.MY_SETTING)  # N: Revealed type is 'builtins.int'

[file mypy_django.ini]
[[mypy_django_plugin]
django_settings = mysettings

[file mysettings.py]
MY_SETTING: int = 1
[/CASE]
