[CASE test_foreign_key_field_with_related_name]
from django.db import models

class Publisher(models.Model):
    pass

class Book(models.Model):
    publisher = models.ForeignKey(to=Publisher, on_delete=models.CASCADE,
                                  related_name='books')

book = Book()
reveal_type(book.publisher)  # E: Revealed type is 'main.Publisher*'

publisher = Publisher()
reveal_type(publisher.books)  # E: Revealed type is 'django.db.models.manager.RelatedManager[main.Book]'

[CASE test_foreign_key_field_creates_attribute_with_underscore_id]
from django.db import models

class Publisher(models.Model):
    pass

class Book(models.Model):
    publisher = models.ForeignKey(to=Publisher, on_delete=models.CASCADE)
    owner = models.ForeignKey(db_column='model_id', to='db.Unknown', on_delete=models.CASCADE)

book = Book()
reveal_type(book.publisher_id)  # E: Revealed type is 'builtins.int'
reveal_type(book.owner_id)  # E: Revealed type is 'Any'

[CASE test_foreign_key_field_different_order_of_params]
from django.db import models

class Publisher(models.Model):
    pass

class Book(models.Model):
    publisher = models.ForeignKey(on_delete=models.CASCADE, to=Publisher,
                                  related_name='books')
    publisher2 = models.ForeignKey(to=Publisher, related_name='books2', on_delete=models.CASCADE)

book = Book()
reveal_type(book.publisher)  # E: Revealed type is 'main.Publisher*'
reveal_type(book.publisher2)  # E: Revealed type is 'main.Publisher*'

publisher = Publisher()
reveal_type(publisher.books)  # E: Revealed type is 'django.db.models.manager.RelatedManager[main.Book]'
reveal_type(publisher.books2)  # E: Revealed type is 'django.db.models.manager.RelatedManager[main.Book]'

[CASE test_to_parameter_as_string_with_application_name__model_imported]
from django.db import models
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from myapp.models import Publisher

class Book(models.Model):
    publisher = models.ForeignKey(to='myapp.Publisher', on_delete=models.CASCADE)

book = Book()
reveal_type(book.publisher)  # E: Revealed type is 'myapp.models.Publisher*'

[file myapp/__init__.py]
[file myapp/models.py]
from django.db import models
class Publisher(models.Model):
    pass

[CASE test_to_parameter_as_string_with_application_name_fallbacks_to_any_if_model_not_present_in_dependency_graph]
from django.db import models

class Book(models.Model):
    publisher = models.ForeignKey(to='myapp.Publisher', on_delete=models.CASCADE)

book = Book()
reveal_type(book.publisher)  # E: Revealed type is 'Any'
reveal_type(book.publisher_id)  # E: Revealed type is 'Any'
Book(publisher_id=1)
Book.objects.create(publisher_id=1)

[file myapp/__init__.py]
[file myapp/models.py]
from django.db import models
class Publisher(models.Model):
    pass

[CASE test_circular_dependency_in_imports_with_foreign_key]
from django.db import models

class App(models.Model):
    def method(self) -> None:
        reveal_type(self.views)  # E: Revealed type is 'django.db.models.manager.RelatedManager[main.View]'
        reveal_type(self.members)  # E: Revealed type is 'django.db.models.manager.RelatedManager[main.Member]'
        reveal_type(self.sheets)  # E: Revealed type is 'django.db.models.manager.RelatedManager[main.Sheet]'
        reveal_type(self.profile)  # E: Revealed type is 'main.Profile'
class View(models.Model):
    app = models.ForeignKey(to=App, related_name='views', on_delete=models.CASCADE)
class Member(models.Model):
    app = models.ForeignKey(related_name='members', on_delete=models.CASCADE, to=App)
class Sheet(models.Model):
    app = models.ForeignKey(App, related_name='sheets', on_delete=models.CASCADE)
class Profile(models.Model):
    app = models.OneToOneField(App, related_name='profile', on_delete=models.CASCADE)

[CASE test_circular_dependency_in_imports_with_string_based]
from django.db import models
from myapp.models import App
class View(models.Model):
    app = models.ForeignKey(to=App, related_name='views', on_delete=models.CASCADE)

reveal_type(View().app.views)  # E: Revealed type is 'django.db.models.manager.RelatedManager[main.View]'
reveal_type(View().app.unknown)
[out]
main:7: error: Revealed type is 'Any'
main:7: error: "App" has no attribute "unknown"

[file myapp/__init__.py]
[file myapp/models.py]
from django.db import models
class App(models.Model):
    def method(self) -> None:
        reveal_type(self.views)  # E: Revealed type is 'django.db.models.manager.RelatedManager[main.View]'

[CASE models_related_managers_work_with_direct_model_inheritance_and_with_inheritance_from_other_model]
from django.db.models import Model
from django.db import models

class App(Model):
    pass

class View(Model):
    app = models.ForeignKey(to=App, on_delete=models.CASCADE, related_name='views')

class View2(View):
    app = models.ForeignKey(to=App, on_delete=models.CASCADE, related_name='views2')

reveal_type(App().views)  # E: Revealed type is 'django.db.models.manager.RelatedManager[main.View]'
reveal_type(App().views2)  # E: Revealed type is 'django.db.models.manager.RelatedManager[main.View2]'
[out]

[CASE models_imported_inside_init_file_foreign_key]
[disable_cache]
from django.db import models
from myapp.models import App
class View(models.Model):
    app = models.ForeignKey(to='myapp.App', related_name='views', on_delete=models.CASCADE)
reveal_type(View().app.views)  # E: Revealed type is 'django.db.models.manager.RelatedManager[main.View]'

[file myapp/__init__.py]
[file myapp/models/__init__.py]
from .app import App
[file myapp/models/app.py]
from django.db import models
class App(models.Model):
    pass
[/CASE]

[CASE models_imported_inside_init_file_one_to_one_field]
[disable_cache]
from django.db import models
from myapp.models import User
class Profile(models.Model):
    user = models.OneToOneField(to='myapp.User', related_name='profile', on_delete=models.CASCADE)
reveal_type(Profile().user.profile)  # E: Revealed type is 'main.Profile'

[file myapp/__init__.py]
[file myapp/models/__init__.py]
from .user import User
[file myapp/models/user.py]
from django.db import models
class User(models.Model):
    pass

[CASE models_triple_circular_reference]
from myapp.models import App
reveal_type(App().owner.profile)  # E: Revealed type is 'myapp.models.profile.Profile'

[file myapp/__init__.py]
[file myapp/models/__init__.py]
from .user import User
from .profile import Profile
from .app import App

[file myapp/models/user.py]
from django.db import models
class User(models.Model):
    pass

[file myapp/models/profile.py]
from django.db import models
from myapp.models import User
class Profile(models.Model):
    user = models.OneToOneField(to='myapp.User', related_name='profile', on_delete=models.CASCADE)

[file myapp/models/app.py]
from django.db import models
class App(models.Model):
    owner = models.ForeignKey(to='myapp.User', on_delete=models.CASCADE, related_name='apps')
[disable_cache]
[/CASE]

[CASE many_to_many_field_converts_to_queryset_of_model_type]
from django.db import models
class App(models.Model):
    pass
class Member(models.Model):
    apps = models.ManyToManyField(to=App, related_name='members')
reveal_type(Member().apps)  # E: Revealed type is 'django.db.models.manager.RelatedManager*[main.App]'
reveal_type(App().members)  # E: Revealed type is 'django.db.models.manager.RelatedManager[main.Member]'
[out]

[CASE many_to_many_works_with_string_if_imported]
from django.db import models
from myapp.models import App
class Member(models.Model):
    apps = models.ManyToManyField(to='myapp.App', related_name='members')
reveal_type(Member().apps)  # E: Revealed type is 'django.db.models.manager.RelatedManager*[myapp.models.App]'

[file myapp/__init__.py]
[file myapp/models.py]
from django.db import models
class App(models.Model):
    pass
[out]

[CASE foreign_key_with_self]
from django.db import models
class User(models.Model):
    parent = models.ForeignKey('self', on_delete=models.CASCADE)
reveal_type(User().parent)  # E: Revealed type is 'main.User*'
[out]

[CASE many_to_many_with_self]
from django.db import models
class User(models.Model):
    friends = models.ManyToManyField('self')
reveal_type(User().friends)  # E: Revealed type is 'django.db.models.manager.RelatedManager*[main.User]'
[out]

[CASE recursively_checking_for_base_model_in_to_parameter]
from django.db import models

class BaseModel(models.Model):
    pass
class ParkingSpot(BaseModel):
    pass
class Booking(BaseModel):
    parking_spot = models.ForeignKey(to=ParkingSpot, null=True, on_delete=models.SET_NULL)
[out]

[CASE if_no_related_name_is_passed_create_default_related_managers]
from django.db import models
class Publisher(models.Model):
    pass
class Book(models.Model):
    publisher = models.ForeignKey(to=Publisher, on_delete=models.CASCADE)
reveal_type(Publisher().book_set)  # E: Revealed type is 'django.db.models.manager.RelatedManager[main.Book]'

[CASE underscore_id_attribute_has_set_type_of_primary_key_if_explicit]
from django.db import models
import datetime
class Publisher(models.Model):
    mypk = models.CharField(max_length=100, primary_key=True)
class Book(models.Model):
    publisher = models.ForeignKey(to=Publisher, on_delete=models.CASCADE)

reveal_type(Book().publisher_id)  # E: Revealed type is 'builtins.str'
Book(publisher_id=1)
Book(publisher_id='hello')
Book(publisher_id=datetime.datetime.now())  # E: Incompatible type for "publisher_id" of "Book" (got "datetime", expected "Union[str, int, Combinable, None]")
Book.objects.create(publisher_id=1)
Book.objects.create(publisher_id='hello')

class Publisher2(models.Model):
    mypk = models.IntegerField(primary_key=True)
class Book2(models.Model):
    publisher = models.ForeignKey(to=Publisher2, on_delete=models.CASCADE)

reveal_type(Book2().publisher_id)  # E: Revealed type is 'builtins.int'
Book2(publisher_id=1)
Book2(publisher_id=[])  # E: Incompatible type for "publisher_id" of "Book2" (got "List[Any]", expected "Union[float, int, str, Combinable, None]")
Book2.objects.create(publisher_id=1)
Book2.objects.create(publisher_id=[])  # E: Incompatible type for "publisher_id" of "Book2" (got "List[Any]", expected "Union[float, int, str, Combinable]")
[out]

[CASE if_model_is_defined_as_name_of_the_class_look_for_it_in_the_same_file]
from django.db import models

class Book(models.Model):
    publisher = models.ForeignKey(to='Publisher', on_delete=models.CASCADE)
class Publisher(models.Model):
    pass
reveal_type(Book().publisher)  # E: Revealed type is 'main.Publisher*'
[out]

[CASE test_foreign_key_field_without_backwards_relation]
from django.db import models

class Publisher(models.Model):
    pass

class Book(models.Model):
    publisher = models.ForeignKey(to=Publisher, on_delete=models.CASCADE,
                                  related_name='+')
    publisher2 = models.ForeignKey(to=Publisher, on_delete=models.CASCADE,
                                   related_name='books2')

book = Book()
reveal_type(book.publisher)  # E: Revealed type is 'main.Publisher*'

publisher = Publisher()
reveal_type(publisher.books)
reveal_type(publisher.books2)  # E: Revealed type is 'django.db.models.manager.RelatedManager[main.Book]'
[out]
main:16: error: Revealed type is 'Any'
main:16: error: "Publisher" has no attribute "books"; maybe "books2"?
[/CASE]

